lib/charms/parca_k8s/v0/parca_scrape.py

```diff
@@ -705,143 +642,68 @@ def __init__(
-        self.topology = ProviderTopology.from_charm(charm)
-
-        self._charm = charm
         self._relation_name = relation_name
         # sanitize job configurations to the supported subset of parameters
         jobs = [] if jobs is None else jobs
         self._jobs = [_sanitize_scrape_configuration(job) for job in jobs]
 
-        events = self._charm.on[self._relation_name]
-        self.framework.observe(events.relation_joined, self._publish_all_relation_data)
-        self.framework.observe(events.relation_changed, self._publish_all_relation_data)
-
-        if not refresh_event:
-            if len(self._charm.meta.containers) == 1:
-                if "kubernetes" in self._charm.meta.series:
-                    # This is a podspec charm
-                    refresh_event = [self._charm.on.update_status]
-                else:
-                    # This is a sidecar/pebble charm
-                    container = list(self._charm.meta.containers.values())[0]
-                    refresh_event = [self._charm.on[container.name.replace("-", "_")].pebble_ready]
-            else:
-                refresh_event = [self._charm.on.update_status]
-
-        else:
-            if not isinstance(refresh_event, list):
-                refresh_event = [refresh_event]
-
-        for ev in refresh_event:
-            self.framework.observe(ev, self._set_unit_ip)
-
-        self.framework.observe(self._charm.on.upgrade_charm, self._publish_all_relation_data)
-        # If there is no leader during relation_joined we will still need to set alert rules.
-        self.framework.observe(self._charm.on.leader_elected, self._publish_all_relation_data)
-
-    def update_scrape_job_spec(self, jobs):
+    def reconcile(self, charm: ops.CharmBase, jobs=None):
         """Update scrape job specification.
 
         This will override the job specs you passed to the constructor.
         Use it if for some reason you can't rely on that being up to date.
         """
-        self._jobs = [_sanitize_scrape_configuration(job) for job in jobs]
-        self._publish_all_relation_data()
+        _validate_relation_by_interface_and_direction(charm, self._relation_name, RELATION_INTERFACE_NAME,
+                                                      RelationRole.provides)
 
-    def _publish_all_relation_data(self, _event=None):
-        self._set_unit_ip()
+        relations = charm.model.relations[self._relation_name]
+        self._publish_unit_data(charm.unit, relations)
 
-        if not self._charm.unit.is_leader():
+        if not charm.unit.is_leader():
             return
 
-        for relation in self._charm.model.relations[self._relation_name]:
-            relation.data[self._charm.app]["scrape_metadata"] = json.dumps(self._scrape_metadata)
-            relation.data[self._charm.app]["scrape_jobs"] = json.dumps(self._scrape_jobs)
+        jobs = [_sanitize_scrape_configuration(job) for job in jobs] or self._jobs or [DEFAULT_JOB]
```

Comment:
I like `_sanitize_scrape_configuration` moving down here because error handling for charms feels better (trying to set a status because you failed to sanitize during init sounds like a mess)

But also, `_sanitize...` is still called in __init__, too, which I think was just a mistake?